keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,849 @@
+  ObjectMeta
+
+  Spec BucketClaimSpec {
+    // Name of the BucketClass
+    BucketClassName string
+  }
+
+  Status BucketClaimStatus {
+    // BucketReady indicates that the bucket is ready for consumpotion
+    // by workloads
+    BucketReady bool
+
+    // BucketName is the name of the provisioned Bucket in response
+    // to this BucketClaim
+    // +optional
+    BucketName string
+  }
 ```
-message ProvisionerGetInfoRequest {
-    // Intentionally left blank
-}
+## BucketClass
+
+Resouce for configuring common properties for multiple Buckets. BucketClass is cluster-scoped.
 
-message ProvisionerGetInfoResponse {
-  // The name MUST follow domain name notation format
-  // (https://tools.ietf.org/html/rfc1035#section-2.3.1). It SHOULD
-  // include the plugin's host company name and the plugin name,
-  // to minimize the possibility of collisions. It MUST be 63
-  // characters or less, beginning and ending with an alphanumeric
-  // character ([a-z0-9A-Z]) with dashes (-), dots (.), and
-  // alphanumerics between. This field is REQUIRED.
-  string name = 1;
+```yaml
+BucketClass {
+  TypeMeta
+  ObjectMeta
+
+  // Provisioner is the name of driver associated with this bucket
+  Provisioner string
+
+  // Protocol is the data API this bucket is expected to adhere to.
+  // The possible values for protocol are:
+  // -  S3: Indicates Amazon S3 protocol
+  // -  Azure: Indicates Microsoft Azure BlobStore protocol
+  // -  GCS: Indicates Google Cloud Storage protocol
+  Protocol Protocol
+
+  // DeletionPolicy is used to specify how COSI should handle deletion of this
+  // bucket. There are 3 possible values:
+  //  - Retain: Indicates that the bucket should not be deleted from the OSP
+  //  - Delete: Indicates that the bucket should be deleted from the OSP
+  //        once all the workloads accessing this bucket are done
+  DeletionPolicy DeletionPolicy
+
+  // Parameters is an opaque map for passing in configuration to a driver
+  // for creating the bucket
+  // +optional
+  Parameters map[string]string
 }
 ```
 
-## ProvisonerCreateBucket
+## BucketAccess
 
-This call is made to create the bucket in the backend. If a bucket that matches both name and parameters already exists, then OK (success) must be returned. If a bucket by same name, but different parameters is provided, then the appropriate error code `ALREADY_EXISTS` must be returned by the provisioner. The call to *ProvisonerCreateBucket* MUST be idempotent.
+A resource to access a Bucket. BucketAccess is namespace-scoped
 
+```yaml
+BucketAccess {
+  TypeMeta
+  ObjectMeta
+
+  Spec BucketAccessSpec {
+    // BucketClaimName is the name of the BucketClaim.
+    // Exactly one of BucketClaimName or BucketName must be set.
+    // +optional
+    BucketClaimName string
+
+    // BucketName is the name of the Bucket for which
+    // credentials need to be generated
+    // Exactly one of BucketClaimName or BucketName must be set.
+    // +optional
+    BucketName string
```

Comment:
I prefer the PVC and Snapshot model where you request a specific resource in the Claim object, and the admin needs to acknowledge it in the admin resource.
- It provides a consistent way between green/brownfield for workloads to refer to and manage Buckets
- It allows for the controller to do basic compatibility checks between what the user requested and what the underlying Bucket can support
- It provides a consistent way for any future policy or quota enforcement to be written